src/client/debugger/jupyter/debuggingManager.ts

```diff
@@ -0,0 +1,213 @@
+
+        this.disposables.push(
+            // track termination of debug sessions
+            vscode.debug.onDidTerminateDebugSession(async (session) => {
+                for (const [doc, dbg] of this.notebookToDebugger.entries()) {
+                    if (dbg && session === (await dbg.session)) {
+                        this.debuggingCellMap.getCellsAnClearQueue(doc);
+                        this.notebookToDebugger.delete(doc);
+                        break;
+                    }
+                }
+            }),
+
+            // track closing of notebooks documents
+            vscode.workspace.onDidCloseNotebookDocument(async (document) => {
+                this.debuggingCellMap.getCellsAnClearQueue(document);
+                const dbg = this.notebookToDebugger.get(document);
+                if (dbg) {
+                    await dbg.stop();
+                }
+                this.fixBreakpoints(document);
+            }),
+
+            // factory for kernel debug adapters
+            vscode.debug.registerDebugAdapterDescriptorFactory('kernel', {
+                createDebugAdapterDescriptor: async (session) => {
+                    const activeDoc = vscode.window.activeNotebookEditor!.document;
+
+                    await this.ensureKernelIsRunning(activeDoc);
+                    const debug = this.getDebuggerByUri(activeDoc);
+
+                    if (debug) {
+                        const notebook = await this.notebookProvider.getOrCreateNotebook({
+                            resource: debug.document.uri,
+                            identity: debug.document.uri,
+                            getOnly: true
+                        });
+                        if (notebook && notebook.session) {
+                            debug.resolve(session);
+                            return new vscode.DebugAdapterInlineImplementation(
+                                new KernelDebugAdapter(session, debug.document, notebook.session, this.debuggingCellMap)
+                            );
+                        } else {
+                            void vscode.window.showInformationMessage('run the kernel');
+                        }
+                    }
+                    // Should not happen, debug sessions should start only from the cell toolbar command
+                    return;
+                }
+            }),
+
+            vscode.commands.registerCommand('jupyter.debugNotebook', () => {
+                const editor = vscode.window.activeNotebookEditor;
+                if (editor) {
+                    void this.startDebugging(editor.document);
+                } else {
+                    void vscode.window.showErrorMessage('No active notebook document to debug');
+                }
+            })
+        );
+    }
+
+    private fixBreakpoints(doc: vscode.NotebookDocument) {
+        const map = new Map<string, vscode.Uri>();
+
+        doc.getCells().forEach((cell, ix) => {
```

Comment:
Minor suggestion, but ix feels like a funny abbreviation. I'd just use index. Makes the loop much more readable to me.

Reply:
cleanup commit